### ISSUE TITLE
Usage on e-Paper Displays (with GxEPD2)

### ISSUE THREAD

**Issue by makuser** [cattanimarco/Ardu-DataVis#22]
I am trying to use your library with GxEPD2 as GFX driver.
Unfortunately, all I am getting is a white/gray screen, no graphs are drawn, nor anything else. Once your class is instantiated and the GFX object is passed to the constructor, all other text that had been on the display before, also disappears. In this case I am not sure if I understand the window parameter properly, eg. if I even should be able to supply my own text next to the graph or not, because I never got it working once.

Do you think you might be able to find out where the issue lies in this case?

**Comment by cattanimarco** [comment]
What type of plot are you trying to draw? Can I have a look at the code? It seems to me that you have have run out of memory (this happens, for example, on the arduino uno if you use the spline datasource)

**Comment by makuser** [comment]
Your basic example code, actually all three/four variations of it.
No memory problems, because it's running on an ESP32.

**Comment by cattanimarco** [comment]
I updated the [wiki](https://github.com/cattanimarco/Grafici-GFX/wiki#window-class) with more info on how the coordinate system works (still a WIP). I hope it helps

**Comment by makuser** [comment]
No, just a white screen, no matter any parameter.

**Comment by cattanimarco** [comment]
At creation Grafici will clear your display. This could be one of your problems. Could you try to comment the `fill_rect` line?
```cpp
	Grafici(Adafruit_GFX &hal)
	    : _display_driver{ hal }
	{
		_display_driver.fill_rect({ 0, 0 }, { 1, 1 }, black);
	}
```
Still, I would expect a black background rather than a white one

**Comment by cattanimarco** [comment]
Can you try this basic example?

```cpp
void setup(void)
{
	gfx.begin();

	Grafici plot{ gfx };

	unsigned int raw_data[] = {1, 2, 4, 5, 2};
	DataArrayXY<unsigned int> data{ raw_data, 5};

	plot.set_color_map(black_and_white);
	plot.line(data.x(), data.y(), data.y(), full_screen);
}

void loop(void)
{
}
```

**Comment by makuser** [comment]
I added the part to initialize GxEPD2 instead of the regular Adafruit GFX, to make it compatible with my display, but I am still only getting a white or rather gray screen.

I also tried putting the actual drawing code info the required nextPage wrapper, which seems like it is required to be able to desigate when to render and actually send the data to the epaper.

**Comment by makuser** [comment]
> At creation Grafici will clear your display. This could be one of your problems. Could you try to comment the `fill_rect` line?
> 
> ```c++
> 	Grafici(Adafruit_GFX &hal)
> 	    : _display_driver{ hal }
> 	{
> 		_display_driver.fill_rect({ 0, 0 }, { 1, 1 }, black);
> 	}
> ```
> 
> Still, I would expect a black background rather than a white one

I already commented that out and before also changed black to white, because I thought it might help to at least display, the other stuff on the screen, but sadly no.

Well, it's a gray/white screen, which looks like an unititialized state, or an aborted reset operation or so. epaper displays special love to display items on the screen.

**Comment by makuser** [comment]
Example code:
```c++
#include <GxEPD2_BW.h>
#define ENABLE_GxEPD2_GFX 0 // also tried turning it on/1

#define DISPLAY_BUSY 4
#define DISPLAY_RST 0
#define DISPLAY_DC 2
#define DISPLAY_CS 15

/* setup display driver */
GxEPD2_BW<GxEPD2_420, GxEPD2_420::HEIGHT> display(GxEPD2_420(DISPLAY_CS, DISPLAY_DC, DISPLAY_RST, DISPLAY_BUSY));
SPIClass hspi(HSPI);

#include "Grafici.h"
constexpr size_t ARRAY_SIZE = 25;
constexpr size_t MAX_VALUE = 255;

unsigned int raw_data[ARRAY_SIZE];
DataArrayXY<unsigned int> data{ raw_data, ARRAY_SIZE, { 0, MAX_VALUE } };

void setup(void)
{
  Serial.begin(115200);
  Serial.println("Hello!");
  /* init display driver and select landscape mode */
  display.init(115200, true, 2, false, hspi, SPISettings(4000000, MSBFIRST, SPI_MODE0));
  Serial.println("step1");

  display.firstPage();
  do {
    display.fillScreen(GxEPD_WHITE);
    display.println("Starting...");
    display.println(millis());
  } while (display.nextPage());

  Serial.println("step1b");

        // Init the plotting library with the display driver
        Grafici plot{ display };
  Serial.println("step2");

        // generate some random data and save it into `raw_data`
        randomSeed(66);
        for (size_t i = 0; i < ARRAY_SIZE; ++i)
        {
                raw_data[i] = random(MAX_VALUE + 1);
        }
  Serial.println("step3");

  auto graph_window = full_screen.sub_window({ 0, 1 }, { .2, .6 });
        // Set the color map
        plot.set_color_map(black_and_white);
  Serial.println("step4");

        // break plot lines in multiple segments to create a gradient of colors between two data points
        PlotLineOpts opts;
        opts.segments = 10;
  Serial.println("step5");

        // Plot data as a line. Color line using `data_y`
        plot.line(data.x(), data.y(), data.y(), graph_window, opts);
  Serial.println("step6");

}

void loop(void)
{
}
```

I also tried extending the do ... while nextPage loop that i mentioned above up to the plot.line part, no change unfortunately.

**Comment by cattanimarco** [comment]
from the docs
> drawing to full screen buffer is done using Adafruit_GFX methods without picture loop or drawCallback
and then calling method display()

try the following:
```cpp
  display.init(115200, true, 2, false, hspi, SPISettings(4000000, MSBFIRST, SPI_MODE0));

  Grafici plot{ display };
  plot.set_color_map(black_and_white);
  plot.line(data.x(), data.y(), data.y(), full_screen);

  display.display(false);
```

Basically, use `display()` rather than `firstPage()` and `nextPage()`

**Comment by cattanimarco** [comment]
I looked at this example
https://github.com/ZinggJM/GxEPD2/blob/master/examples/GxEPD2_NotPagedExample/GxEPD2_NotPagedExample.ino

**Comment by makuser** [comment]
> from the docs
> 
> > drawing to full screen buffer is done using Adafruit_GFX methods without picture loop or drawCallback
> > and then calling method display()
> 
> Basically, use `display()` rather than `firstPage()` and `nextPage()`

Thank you very much, using a non-paged approach for the views where your graph library is used, allowed me to display the graph on the e-Paper screen.

**Comment by makuser** [comment]
> ```c++
> 	Grafici(Adafruit_GFX &hal)
> 	    : _display_driver{ hal }
> 	{
> 		_display_driver.fill_rect({ 0, 0 }, { 1, 1 }, black);
> 	}
> ```

Being able to disable this, would be awesome btw ;)

**Comment by cattanimarco** [comment]
done

**Comment by makuser** [comment]
> done

Well done! 😄  Thank you again!

May this be released as 3.1.1 or 3.2.0?
